code/modules/events/crystal_event.dm

```diff
@@ -0,0 +1,408 @@
+GLOBAL_LIST_INIT(small_wave, list(/obj/structure/crystal_portal/small=4, /obj/structure/crystal_portal/medium=1)) //5
+GLOBAL_LIST_INIT(medium_wave, list(/obj/structure/crystal_portal/small=4, /obj/structure/crystal_portal/medium=3, /obj/structure/crystal_portal/big=1)) //8
+GLOBAL_LIST_INIT(big_wave, list(/obj/structure/crystal_portal/small=5, /obj/structure/crystal_portal/medium=3, /obj/structure/crystal_portal/big=2, /obj/structure/crystal_portal/huge=1)) //11
+GLOBAL_LIST_INIT(huge_wave, list(/obj/structure/crystal_portal/small=7, /obj/structure/crystal_portal/medium=5, /obj/structure/crystal_portal/big=3, /obj/structure/crystal_portal/huge=2)) //17
+GLOBAL_LIST_EMPTY(crystal_portals)
+/*
+This section is for the event controller
+*/
+/datum/round_event_control/crystal_invasion
+	name = "Crystal Invasion"
+	typepath = /datum/round_event/crystal_invasion
+	weight = 2
+	min_players = 35
+	max_occurrences = 1
+	earliest_start = 45 MINUTES
+
+/datum/round_event/crystal_invasion
+	announceWhen = 1
+	///Stores the type of the wave for the event
+	var/list/wave_type
+	///Is the name of the wave, used to check wich wave will be generated
+	var/wave_name
+	///Max number of portals that can spawn per type of wave
+	var/portal_numbers
+
+/datum/round_event/crystal_invasion/announce()
+	priority_announce("WARNING - Destabilization of the Supermatter Crystal Matrix detected, please stand by waiting further instructions", "Alert", 'sound/misc/notice1.ogg')
+
+/datum/round_event/crystal_invasion/New()
+	. = ..()
+	if(!wave_type)
+		choose_wave_type()
+
+///Choose the type of the wave
+/datum/round_event/crystal_invasion/proc/choose_wave_type()
+	if(!wave_name)
+		wave_name = pickweight(list(
+			"small wave" = 60,
+			"medium wave" = 30,
+			"big wave" = 7,
+			"huge wave" = 3))
+	switch(wave_name)
+		if("small wave")
+			wave_type = GLOB.small_wave
+			portal_numbers = pick(2, 3, 4, 5)
+		if("medium wave")
+			wave_type = GLOB.medium_wave
+			portal_numbers = pick(4, 5, 6, 7)
+		if("big wave")
+			wave_type = GLOB.big_wave
+			portal_numbers = pick(6, 7, 8, 9)
+		if("huge wave")
+			wave_type = GLOB.huge_wave
+			portal_numbers = pick(9, 10, 11, 12)
+		else
+			WARNING("Wave name of [wave_name] not recognised.")
+			kill()
+
+	sleep(10 SECONDS)
```

Comment:
how does callback works?

Reply:
Looks like this `addtimer(CALLBACK(object_to_call_on, .proc_you_want_to_call, arguments), time in deciseconds)`
Instead of relying on byond to do the sleep/recall for us, it uses the timer system we've built, which tends to be a bit safer on the overall. Not sure if this is a good usecase or not, but it should help prevent unintended issues in future.

Reply:
done